src/custom/float.rs

```diff
@@ -89,20 +94,83 @@ impl<'c, H: HugrView> CodegenExtension<'c, H> for FloatOpsCodegenExtension {
     }
 }
 
-// we allow dead code for now, but once we implement the emitter, we should
-// remove this
-#[allow(dead_code)]
+/// An emitter for [hugr::std_extensions::arithmetic::float_ops] ops.
 struct FloatOpEmitter<'c, 'd, H>(&'d mut EmitFuncContext<'c, H>);
 
+/// Emit a float comparison operation.
+fn emit_fcmp<'c, H: HugrView>(
+    context: &mut EmitFuncContext<'c, H>,
+    args: EmitOpArgs<'c, CustomOp, H>,
+    pred: inkwell::FloatPredicate,
+) -> Result<()> {
+    let true_val = emit_value(context, &Value::true_val())?;
+    let false_val = emit_value(context, &Value::false_val())?;
+
+    emit_custom_binary_op(context, args, |builder, (lhs, rhs), _| {
+        // get result as an i1
+        let r = builder.build_float_compare(
+            pred,
+            lhs.into_float_value(),
+            rhs.into_float_value(),
+            "",
+        )?;
+        // convert to whatever BOOL_T is
+        Ok(vec![builder.build_select(r, true_val, false_val, "")?])
+    })
+}
+
 impl<'c, H: HugrView> EmitOp<'c, CustomOp, H> for FloatOpEmitter<'c, '_, H> {
     fn emit(&mut self, args: EmitOpArgs<'c, CustomOp, H>) -> Result<()> {
-        use hugr::ops::NamedOp;
-        let name = args.node().name();
-        // This looks strange now, but we will add cases for ops piecemeal, as
-        // in the analgous match expression in `IntOpEmitter`.
-        #[allow(clippy::match_single_binding)]
-        match name.as_str() {
-            n => Err(anyhow!("FloatOpEmitter: unknown op: {n}")),
+        let op = FloatOps::from_optype(&args.node().generalise()).ok_or(anyhow!(
+            "LogicOpEmitter: from_optype_failed: {:?}",
+            args.node().as_ref()
+        ))?;
+        // We emit the float comparison variants where NaN is an absorbing value.
+        // Any comparison with NaN is always false.
```

Comment:
Just checking, does this match the behaviour in float constant folding?

Reply:
We use rust's float semantics, which compile to the "O" variants used here.
https://github.com/CQCL/hugr/blob/06df427dcb6538b2b5dcd931e7d7a57ed3193f31/hugr-core/src/std_extensions/arithmetic/float_ops/const_fold.rs#L73-L78

Playground:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=abd2915b47e9f045fa9361bba729246d